File: Problems/0122_Best_Time_to_Buy_and_Sell_Stock2/Project_CS/Best_Time_to_Buy_and_Sell_Stock2.cs
using System;

public class Solution
{
    public int MaxProfit(int[] prices)
    {
        int max = 0;
        int sum_max = 0;
        int min = int.MaxValue;

        for(int i = 0; i < prices.Length; i++)
        {
            if(min > prices[i])
                min = prices[i];
            if(prices[i] - min > max)
                max = prices[i] - min;
            if(prices[i] - min > 0) {
                sum_max += prices[i] - min;
                min = prices[i];
            }
        }
        
        if (max > sum_max)
            return max;
        else
            return sum_max;
    }

    public int[] str_to_int_array(string s)
    {
        if (s.Length <= 0)
            return null;

        string[] flds = s.Split(',');
        int[] nums = new int[flds.Length];

        if (flds.Length <= 0)
            return nums;

        for (int i = 0; i < nums.Length; ++i)
        {
            nums[i] = int.Parse(flds[i]);
        }

        return nums;
    }

    public string output_int_array(int[] nums)
    {
        if (nums.Length <= 0)
            return "";

        string resultStr = "[" +  nums[0].ToString();
 
        for (int i = 1; i < nums.Length; ++i)
        {
            resultStr += "," + nums[i].ToString();
        }

        return resultStr + "]";
    }

    public void Main(string args)
    {
        string flds = args.Replace("\"", "").Replace(" ", "").Replace("[", "").Replace("]", "");
        int[] prices = str_to_int_array(flds);
        Console.WriteLine("prices[] = " + output_int_array(prices));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        int result = MaxProfit(prices);

        sw.Stop();
        Console.WriteLine("Result = " + result.ToString());
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0100_0199/0122_Best_Time_to_Buy_and_Sell_Stock2/Project_Python3/Best_Time_to_Buy_and_Sell_Stock2.py
# coding: utf-8

import os
import sys
import time

class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        max = 0
        sum_max = 0
        min = sys.maxsize
        for i in range(len(prices)):
            if min > prices[i]:
                min = prices[i]
            profit = prices[i] - min
            if profit > max:
                max = profit
            if profit > 0:
                sum_max += profit
                min = prices[i]
        
        if max > sum_max:
            return max
        else:
            return sum_max

def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    flds = temp.replace("\"","").replace("[","").replace("]","").rstrip()

    prices = [int(val) for val in flds.split(",")]
    print("prices = %s" %prices)

    time0 = time.time()

    sl = Solution()
    result = sl.maxProfit(prices)

    time1 = time.time()

    print("result = %s" %result)
    print("Execute time ... : %f[s]\n" %(time1 - time0))

if __name__ == "__main__":
    main()
